page_decoder.go

```diff
@@ -0,0 +1,399 @@
+package parquet
+
+import (
+	"fmt"
+	"io"
+	"sync"
+	"unsafe"
+
+	"github.com/segmentio/parquet-go/encoding"
+	"github.com/segmentio/parquet-go/encoding/plain"
+	"github.com/segmentio/parquet-go/internal/bits"
+)
+
+type filePageDecoder struct {
+	remain             int
+	numValues          int
+	maxRepetitionLevel int8
+	maxDefinitionLevel int8
+	columnIndex        int16
+	repetitions        levelDecoder
+	definitions        levelDecoder
+	values             PageValues
+}
+
+func newFilePageDecoder(numValues int, columnIndex int16, maxRepetitionLevel, maxDefinitionLevel int8, repetitions, definitions encoding.Decoder, values PageValues) *filePageDecoder {
+	if repetitions != nil {
+		repetitions.SetBitWidth(bits.Len8(maxRepetitionLevel))
+	}
+	if definitions != nil {
+		definitions.SetBitWidth(bits.Len8(maxDefinitionLevel))
+	}
+	return &filePageDecoder{
+		remain:             numValues,
+		numValues:          numValues,
+		maxRepetitionLevel: maxRepetitionLevel,
+		maxDefinitionLevel: maxDefinitionLevel,
+		columnIndex:        ^columnIndex,
+		repetitions:        makeLevelDecoder(repetitions),
+		definitions:        makeLevelDecoder(definitions),
+		values:             values,
+	}
+}
+
+func (r *filePageDecoder) Close() error {
+	r.remain = 0
+	r.repetitions.close()
+	r.definitions.close()
+	return r.values.Close()
+}
+
+func (r *filePageDecoder) ReadValues(values []Value) (int, error) {
+	read := 0
+
+	for r.remain > 0 && len(values) > 0 {
+		var err error
+		var repetitionLevels []int8
+		var definitionLevels []int8
+		var numValues = r.remain
+
+		if len(values) < numValues {
+			numValues = len(values)
+		}
+
+		if r.maxRepetitionLevel > 0 {
+			repetitionLevels, err = r.repetitions.peekLevels()
+			if err != nil {
+				return read, fmt.Errorf("decoding repetition level from data page of column %d: %w", ^r.columnIndex, err)
+			}
+			if len(repetitionLevels) < numValues {
+				numValues = len(repetitionLevels)
+			}
+		}
+
+		if r.maxDefinitionLevel > 0 {
+			definitionLevels, err = r.definitions.peekLevels()
+			if err != nil {
+				return read, fmt.Errorf("decoding definition level from data page of column %d: %w", ^r.columnIndex, err)
```

Comment:
this bit flipping is a trip every time I read it. wondering if we should just make this a function that would get inlined? 

make the variable name `_columnIndex` and then `r.columnIndex()` returns the value or something?

Reply:
Like if I had to work with this codebase I guarantee I would make a mistake while using columnIndex by not doing the right thing.

Reply:
I agree this is tricky to get right, I'll revisit 👍